File: Packages/TeamTasksKit/Sources/TeamTasksCore/Logic/DueBucket.swift
import Foundation

/// Due-date section of a task list ("Mes tâches"), in display order.
public enum DueBucket: Int, Sendable, Hashable, Codable, CaseIterable, Comparable, Identifiable {
    /// Not done, due strictly before now.
    case overdue
    /// Not done, due from now until the end of today.
    case today
    /// Not done, due after today and before the start of next week (per the calendar's `firstWeekday`).
    case thisWeek
    /// Not done, due from the start of next week on.
    case later
    /// Not done, without due date.
    case noDueDate
    /// Done tasks, whatever their due date.
    case done

    public var id: Int { rawValue }

    /// French section title.
    public var title: String {
        switch self {
        case .overdue: "En retard"
        case .today: "Aujourd'hui"
        case .thisWeek: "Cette semaine"
        case .later: "Plus tard"
        case .noDueDate: "Sans échéance"
        case .done: "Terminées"
        }
    }

    public static func < (lhs: DueBucket, rhs: DueBucket) -> Bool { lhs.rawValue < rhs.rawValue }

    /// Bucket of one task. Build a `DueBucketBoundaries` once when classifying many tasks.
    public static func bucket(for task: TaskItem, now: Date, calendar: Calendar) -> DueBucket {
        DueBucketBoundaries(now: now, calendar: calendar).bucket(for: task)
    }

    /// Groups tasks into ordered, non-empty sections.
    /// - Parameters:
    ///   - sort: order inside the not-done sections (default: due date, then priority, then title).
    ///     The "Terminées" section lists the most recently completed first, then follows `sort`.
    public static func sections(
        for tasks: [TaskItem],
        now: Date,
        calendar: Calendar,
        sort: TaskSort = .dueDate
    ) -> [DueSection] {
        DueBucketBoundaries(now: now, calendar: calendar).sections(for: tasks, sort: sort)
    }
}

/// A non-empty section of tasks sharing a `DueBucket`.
public struct DueSection: Sendable, Hashable, Identifiable {
    public var bucket: DueBucket
    public var tasks: [TaskItem]

    public var id: DueBucket { bucket }
    public var title: String { bucket.title }

    public init(bucket: DueBucket, tasks: [TaskItem]) {
        self.bucket = bucket
        self.tasks = tasks
    }
}

/// Day boundaries used to classify due dates, computed once from `now` and the injected calendar
/// (time zone and first weekday). Correct across daylight-saving transitions: boundaries are calendar
/// midnights, not multiples of 24 hours.
public struct DueBucketBoundaries: Sendable, Hashable {
    public let now: Date
    /// First instant of today.
    public let startOfToday: Date
    /// First instant of tomorrow (end of the "Aujourd'hui" bucket).
    public let startOfTomorrow: Date
    /// First instant of next week (end of the "Cette semaine" bucket). Equals `startOfTomorrow`
    /// on the last day of the week (Sunday with French conventions), making "Cette semaine" empty.
    public let startOfNextWeek: Date

    public init(now: Date, calendar: Calendar) {
        self.now = now
        let startOfToday = calendar.startOfDay(for: now)
        self.startOfToday = startOfToday

        let weekday = calendar.component(.weekday, from: startOfToday)
        let daysSinceWeekStart = ((weekday - calendar.firstWeekday) % 7 + 7) % 7
        startOfTomorrow = Self.startOfDay(daysAfter: 1, startOfToday, calendar)
        startOfNextWeek = Self.startOfDay(daysAfter: 7 - daysSinceWeekStart, startOfToday, calendar)
    }

    public func bucket(for task: TaskItem) -> DueBucket {
        if task.status == .done { return .done }
        guard let dueAt = task.dueAt else { return .noDueDate }
        if dueAt < now { return .overdue }
        if dueAt < startOfTomorrow { return .today }
        if dueAt < startOfNextWeek { return .thisWeek }
        return .later
    }

    /// See `DueBucket.sections(for:now:calendar:sort:)`.
    public func sections(for tasks: [TaskItem], sort: TaskSort = .dueDate) -> [DueSection] {
        var grouped: [DueBucket: [TaskItem]] = [:]
        for task in tasks {
            grouped[bucket(for: task), default: []].append(task)
        }
        return DueBucket.allCases.compactMap { bucket in
            guard let tasks = grouped[bucket], !tasks.isEmpty else { return nil }
            let ordered = bucket == .done ? Self.sortDone(tasks, sort: sort) : sort.sorted(tasks)
            return DueSection(bucket: bucket, tasks: ordered)
        }
    }

    /// Most recently completed first (unknown completion date last), ties broken by `sort`.
    private static func sortDone(_ tasks: [TaskItem], sort: TaskSort) -> [TaskItem] {
        let base = sort.sorted(tasks)
        let keyed = base.enumerated().map { (offset: $0.offset, task: $0.element) }
        return keyed.sorted { lhs, rhs in
            switch (lhs.task.completedAt, rhs.task.completedAt) {
            case let (l?, r?) where l != r: return l > r
            case (_?, nil): return true
            case (nil, _?): return false
            default: return lhs.offset < rhs.offset
            }
        }.map { $0.task }
    }

    private static func startOfDay(daysAfter days: Int, _ startOfToday: Date, _ calendar: Calendar) -> Date {
        let shifted = calendar.date(byAdding: .day, value: days, to: startOfToday)
            ?? startOfToday.addingTimeInterval(Double(days) * 86_400)
        return calendar.startOfDay(for: shifted)
    }
}

File: android/core/src/main/kotlin/io/github/notkanaa/equipe/core/logic/DueBucket.kt
package io.github.notkanaa.equipe.core.logic

import io.github.notkanaa.equipe.core.AppCalendar
import io.github.notkanaa.equipe.core.TaskItem
import io.github.notkanaa.equipe.core.TaskStatus
import java.time.DayOfWeek
import java.time.Instant

// Port of TeamTasksCore/Logic/DueBucket.swift.

/** Due-date section of a task list ("Mes tâches"), in display order (the natural enum order). */
enum class DueBucket(val title: String) {
    /** Not done, due strictly before now. */
    OVERDUE("En retard"),

    /** Not done, due from now until the end of today. */
    TODAY("Aujourd’hui"),

    /** Not done, due after today and before the start of next week (per the first day of the week). */
    THIS_WEEK("Cette semaine"),

    /** Not done, due from the start of next week on. */
    LATER("Plus tard"),

    /** Not done, without due date. */
    NO_DUE_DATE("Sans échéance"),

    /** Done tasks, whatever their due date. */
    DONE("Terminées"),
    ;

    /** Swift raw value (display rank). */
    val rawValue: Int get() = ordinal

    val id: Int get() = ordinal

    companion object {
        /**
         * Bucket of one task. Build a [DueBucketBoundaries] once when classifying many tasks.
         * @param firstDayOfWeek defaults to the calendar's (Monday with French conventions).
         */
        fun bucket(
            task: TaskItem,
            now: Instant,
            calendar: AppCalendar,
            firstDayOfWeek: DayOfWeek = calendar.weekFields.firstDayOfWeek,
        ): DueBucket = DueBucketBoundaries(now, calendar, firstDayOfWeek).bucket(task)

        /**
         * Groups tasks into ordered, non-empty sections.
         * @param sort order inside the not-done sections (default: due date, then priority, then title).
         *   The "Terminées" section lists the most recently completed first, then follows [sort].
         */
        fun sections(
            tasks: List<TaskItem>,
            now: Instant,
            calendar: AppCalendar,
            sort: TaskSort = TaskSort.DUE_DATE,
            firstDayOfWeek: DayOfWeek = calendar.weekFields.firstDayOfWeek,
        ): List<DueSection> = DueBucketBoundaries(now, calendar, firstDayOfWeek).sections(tasks, sort)
    }
}

/** A non-empty section of tasks sharing a [DueBucket]. */
data class DueSection(
    val bucket: DueBucket,
    val tasks: List<TaskItem>,
) {
    val id: DueBucket get() = bucket
    val title: String get() = bucket.title
}

/**
 * Day boundaries used to classify due dates, computed once from [now] and the injected calendar (time zone) and first
 * day of the week. Correct across daylight-saving transitions: boundaries are calendar midnights (the first instant of
 * the day when midnight is skipped), not multiples of 24 hours.
 *
 * @param firstDayOfWeek defaults to the calendar's (Monday with French conventions); Swift reads it from the injected
 *   `Calendar.firstWeekday`.
 */
class DueBucketBoundaries(
    val now: Instant,
    calendar: AppCalendar,
    firstDayOfWeek: DayOfWeek = calendar.weekFields.firstDayOfWeek,
) {
    /** First instant of today. */
    val startOfToday: Instant

    /** First instant of tomorrow (end of the "Aujourd’hui" bucket). */
    val startOfTomorrow: Instant

    /**
     * First instant of next week (end of the "Cette semaine" bucket). Equals [startOfTomorrow] on the last day of the
     * week (Sunday with French conventions), making "Cette semaine" empty.
     */
    val startOfNextWeek: Instant

    init {
        val today = calendar.localDate(now)
        startOfToday = calendar.startOfDay(today)
        val daysSinceWeekStart = Math.floorMod(today.dayOfWeek.value - firstDayOfWeek.value, 7)
        startOfTomorrow = calendar.startOfDay(today.plusDays(1))
        startOfNextWeek = calendar.startOfDay(today.plusDays((7 - daysSinceWeekStart).toLong()))
    }

    fun bucket(task: TaskItem): DueBucket {
        if (task.status == TaskStatus.DONE) return DueBucket.DONE
        val dueAt = task.dueAt ?: return DueBucket.NO_DUE_DATE
        return when {
            dueAt < now -> DueBucket.OVERDUE
            dueAt < startOfTomorrow -> DueBucket.TODAY
            dueAt < startOfNextWeek -> DueBucket.THIS_WEEK
            else -> DueBucket.LATER
        }
    }

    /** See [DueBucket.sections]. */
    fun sections(tasks: List<TaskItem>, sort: TaskSort = TaskSort.DUE_DATE): List<DueSection> {
        val grouped = HashMap<DueBucket, MutableList<TaskItem>>()
        for (task in tasks) {
            grouped.getOrPut(bucket(task)) { ArrayList() }.add(task)
        }
        return DueBucket.entries.mapNotNull { bucket ->
            val bucketTasks = grouped[bucket]
            if (bucketTasks.isNullOrEmpty()) return@mapNotNull null
            val ordered = if (bucket == DueBucket.DONE) sortDone(bucketTasks, sort) else sort.sorted(bucketTasks)
            DueSection(bucket, ordered)
        }
    }

    override fun equals(other: Any?): Boolean =
        other is DueBucketBoundaries && now == other.now && startOfToday == other.startOfToday &&
            startOfTomorrow == other.startOfTomorrow && startOfNextWeek == other.startOfNextWeek

    override fun hashCode(): Int {
        var result = now.hashCode()
        result = 31 * result + startOfToday.hashCode()
        result = 31 * result + startOfTomorrow.hashCode()
        result = 31 * result + startOfNextWeek.hashCode()
        return result
    }

    override fun toString(): String =
        "DueBucketBoundaries(now=$now, startOfToday=$startOfToday, startOfTomorrow=$startOfTomorrow, " +
            "startOfNextWeek=$startOfNextWeek)"

    private companion object {
        /** Most recently completed first (unknown completion date last), ties broken by [sort] (stable sort). */
        fun sortDone(tasks: List<TaskItem>, sort: TaskSort): List<TaskItem> =
            sort.sorted(tasks).sortedWith { lhs, rhs ->
                val left = lhs.completedAt
                val right = rhs.completedAt
                when {
                    left != null && right != null -> right.compareTo(left)
                    left != null -> -1
                    right != null -> 1
                    else -> 0
                }
            }
    }
}
